File: app/src/main/java/com/example/yakuzaiapp/util/GtinNormalizer.kt
package com.example.yakuzaiapp.util

/**
 * バーコード文字列から GTIN-14 を抽出する。
 *
 * - AI=01 プレフィックス付きの 16 桁は先頭 14 桁を返す
 * - 既に GTIN-14 の場合はそのまま返す
 * - GTIN-13 の場合は先頭に 0 を付けて GTIN-14 にする
 * - その他は null を返す
 */
fun normalizeGtin(rawCode: String): String? {
    val value = rawCode
        .map { ch ->
            when (ch) {
                in '０'..'９' -> '0' + (ch - '０')
                else -> ch
            }
        }
        .joinToString("")
        .filter(Char::isDigit)
    val candidate = when {
        value.startsWith("01") && value.length >= 16 -> value.substring(2, 16)
        value.length == 14 -> value
        value.length == 13 -> "0$value"
        else -> null
    }
    return candidate?.takeIf(::hasValidGtinCheckDigit)
}

/**
 * マスター照合用のバーコードを正規化する。
 *
 * 市販品のGTINはGTIN-14へ正規化し、院内製剤や材料で使われる施設独自コードは
 * 読み取った文字列をそのまま照合キーとして扱う。
 */
fun normalizeMasterBarcode(rawCode: String): String? {
    normalizeGtin(rawCode)?.let { return it }
    val value = rawCode.trim()
    if (value.length !in 3..64 || value.any(Char::isISOControl)) return null

    val digits = value.filter(Char::isDigit)
    val looksLikeInvalidGtin = value.all { it.isDigit() || it.isWhitespace() || it == '(' || it == ')' } &&
        (digits.length == 13 || digits.length == 14 || (digits.startsWith("01") && digits.length >= 16))
    if (looksLikeInvalidGtin) return null

    return value.takeIf { digits.isNotEmpty() }
}

private fun hasValidGtinCheckDigit(gtin: String): Boolean {
    if (gtin.isEmpty() || !gtin.all(Char::isDigit)) return false
    val checkDigit = gtin.last().digitToInt()
    var sum = 0
    var weight = 3
    for (i in gtin.length - 2 downTo 0) {
        sum += gtin[i].digitToInt() * weight
        weight = if (weight == 3) 1 else 3
    }
    val expected = (10 - (sum % 10)) % 10
    return checkDigit == expected
}

/**
 * GTIN-14 のうち、1〜8 で始まる包装・箱バーコードかどうかを判定する。
 *
 * - 14 桁のみを対象とする
 * - 先頭 0 はシート系、9 は予備コードとして false
 */
fun isPackageBarcode(gtin: String): Boolean {
    val value = gtin.trim()
    return value.length == 14 && value.all(Char::isDigit) && value.first() in '1'..'8'
}

File: ios-yakupita/Sources/YakupitaCore/GtinNormalizer.swift
import Foundation

public enum GtinNormalizer {
    public static func normalize(_ rawCode: String) -> String? {
        let digits = normalizedDigits(in: rawCode)
        let candidate: String

        if digits.hasPrefix("01"), digits.count >= 16 {
            candidate = substring(digits, from: 2, length: 14)
        } else if digits.count == 14 {
            candidate = digits
        } else if digits.count == 13 {
            candidate = "0" + digits
        } else {
            return nil
        }

        return hasValidCheckDigit(candidate) ? candidate : nil
    }

    public static func normalizeMasterBarcode(_ rawCode: String) -> String? {
        if let gtin = normalize(rawCode) {
            return gtin
        }

        let value = rawCode.trimmingCharacters(in: .whitespacesAndNewlines)
        guard (3...64).contains(value.count), !containsControlCharacter(value) else {
            return nil
        }

        let decimalDigits = value.filter(\.isWholeNumber)
        let containsOnlyGtinFormatting = value.allSatisfy {
            $0.isWholeNumber || $0.isWhitespace || $0 == "(" || $0 == ")"
        }
        let looksLikeInvalidGtin = containsOnlyGtinFormatting && (
            decimalDigits.count == 13 ||
                decimalDigits.count == 14 ||
                (decimalDigits.hasPrefix("01") && decimalDigits.count >= 16)
        )

        guard !looksLikeInvalidGtin, !decimalDigits.isEmpty else {
            return nil
        }
        return value
    }

    public static func isPackageBarcode(_ gtin: String) -> Bool {
        let value = gtin.trimmingCharacters(in: .whitespacesAndNewlines)
        guard value.count == 14, value.allSatisfy({ $0.isASCII && $0.isWholeNumber }) else {
            return false
        }
        guard let firstValue = value.first?.asciiValue else { return false }
        return (0x31...0x38).contains(firstValue)
    }

    private static func normalizedDigits(in rawCode: String) -> String {
        var result = ""
        result.reserveCapacity(rawCode.count)

        for scalar in rawCode.unicodeScalars {
            switch scalar.value {
            case 0x30...0x39:
                result.unicodeScalars.append(scalar)
            case 0xFF10...0xFF19:
                guard let ascii = UnicodeScalar(0x30 + scalar.value - 0xFF10) else { continue }
                result.unicodeScalars.append(ascii)
            default:
                continue
            }
        }
        return result
    }

    private static func hasValidCheckDigit(_ gtin: String) -> Bool {
        guard !gtin.isEmpty, gtin.allSatisfy({ $0.isASCII && $0.isWholeNumber }) else {
            return false
        }

        let digits = gtin.compactMap(\.wholeNumberValue)
        guard digits.count == gtin.count, let checkDigit = digits.last else { return false }

        var sum = 0
        var weight = 3
        for digit in digits.dropLast().reversed() {
            sum += digit * weight
            weight = weight == 3 ? 1 : 3
        }
        return checkDigit == (10 - (sum % 10)) % 10
    }

    private static func containsControlCharacter(_ value: String) -> Bool {
        value.unicodeScalars.contains { scalar in
            scalar.properties.generalCategory == .control
        }
    }

    private static func substring(_ value: String, from offset: Int, length: Int) -> String {
        let start = value.index(value.startIndex, offsetBy: offset)
        let end = value.index(start, offsetBy: length)
        return String(value[start..<end])
    }
}
